# CSE 482 Project — Part 1: Data Collection and Preprocessing

**Project topic:** Movie Recommendation System using Collaborative Filtering  
**Dataset:** MovieLens `ml-latest-small`  
**Student:** `Mehrshad Bagherebadian`  
**Team:** Individual project

This notebook supports **Part 1: Data Collection and Preprocessing** of the CSE 482 recommender system project.

The final project requires a recommender system that applies:
1. User-based collaborative filtering
2. Item-based collaborative filtering
3. Training/test evaluation
4. A web-based interface

This Part 1 notebook focuses only on the data stage:
- loading the MovieLens files,
- checking data quality,
- preprocessing the data,
- preparing modeling-ready tables,
- storing the cleaned data in a local MySQL database.

## 1. Project Folder Structure

This notebook assumes the following folder structure:

```text
CSE482_Project/
├── dataset/
│   ├── ratings.csv
│   ├── movies.csv
│   ├── tags.csv
│   ├── links.csv
│   └── README.txt
├── reports/
├── notebooks/
│   └── Part1_Data_Preprocessing.ipynb
└── database/
```

Because this notebook is saved inside the `notebooks/` folder, dataset files are accessed using paths like:

```python
../dataset/ratings.csv
```

In [2]:
from pathlib import Path
import pandas as pd
import numpy as np

current_dir = Path.cwd()

if current_dir.name.lower() == "notebooks":
    PROJECT_ROOT = current_dir.parent
else:
    
    PROJECT_ROOT = current_dir

DATA_DIR = PROJECT_ROOT / "dataset"
REPORTS_DIR = PROJECT_ROOT / "reports"
DATABASE_DIR = PROJECT_ROOT / "database"

REPORTS_DIR.mkdir(exist_ok=True)
DATABASE_DIR.mkdir(exist_ok=True)

print("Project root:", PROJECT_ROOT)
print("Dataset folder:", DATA_DIR)
print("Reports folder:", REPORTS_DIR)
print("Database folder:", DATABASE_DIR)

Project root: /Users/mehrshad/Documents/CSE482_Project
Dataset folder: /Users/mehrshad/Documents/CSE482_Project/dataset
Reports folder: /Users/mehrshad/Documents/CSE482_Project/reports
Database folder: /Users/mehrshad/Documents/CSE482_Project/database


## 2. Load the Dataset

The MovieLens latest-small dataset contains four CSV files:

| File | Purpose |
|---|---|
| `ratings.csv` | User-movie rating records |
| `movies.csv` | Movie titles and genre metadata |
| `tags.csv` | User-generated text tags |
| `links.csv` | External IDs for IMDb and TMDb |

For the core recommender system, the two most important files are:
- `ratings.csv`
- `movies.csv`

The `tags.csv` and `links.csv` files are inspected for completeness, but they are not required for the basic collaborative filtering model.

In [3]:
ratings = pd.read_csv(DATA_DIR / "ratings.csv")
movies = pd.read_csv(DATA_DIR / "movies.csv")
tags = pd.read_csv(DATA_DIR / "tags.csv")
links = pd.read_csv(DATA_DIR / "links.csv")

print("ratings shape:", ratings.shape)
print("movies shape:", movies.shape)
print("tags shape:", tags.shape)
print("links shape:", links.shape)

ratings shape: (100836, 4)
movies shape: (9742, 3)
tags shape: (3683, 4)
links shape: (9742, 3)


In [4]:
display(ratings.head())
display(movies.head())
display(tags.head())
display(links.head())

,userId,movieId,rating,timestamp
0,1,1,4.0,964982703
1,1,3,4.0,964981247
2,1,6,4.0,964982224
3,1,47,5.0,964983815
4,1,50,5.0,964982931


,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


,userId,movieId,tag,timestamp
0,2,60756,funny,1445714994
1,2,60756,Highly quotable,1445714996
2,2,60756,will ferrell,1445714992
3,2,89774,Boxing story,1445715207
4,2,89774,MMA,1445715200


,movieId,imdbId,tmdbId
0,1,114709,862.0
1,2,113497,8844.0
2,3,113228,15602.0
3,4,114885,31357.0
4,5,113041,11862.0


## 3. Dataset Size and Basic Description

According to the MovieLens README, the latest-small dataset contains:
- 100,836 ratings
- 3,683 tag applications
- 9,742 movies
- 610 users
- ratings created between March 29, 1996 and September 24, 2018

The following cells verify these numbers directly from the CSV files.

In [5]:
dataset_summary = pd.DataFrame({
    "Dataset": ["ratings", "movies", "tags", "links"],
    "Rows": [len(ratings), len(movies), len(tags), len(links)],
    "Columns": [ratings.shape[1], movies.shape[1], tags.shape[1], links.shape[1]]
})

display(dataset_summary)

print("Unique users in ratings:", ratings["userId"].nunique())
print("Unique movies in ratings:", ratings["movieId"].nunique())
print("Unique movies in movies.csv:", movies["movieId"].nunique())
print("Unique tag values:", tags["tag"].nunique())

,Dataset,Rows,Columns
0,ratings,100836,4
1,movies,9742,3
2,tags,3683,4
3,links,9742,3


Unique users in ratings: 610
Unique movies in ratings: 9724
Unique movies in movies.csv: 9742
Unique tag values: 1589


In [6]:
ratings_time_min = pd.to_datetime(ratings["timestamp"], unit="s").min()
ratings_time_max = pd.to_datetime(ratings["timestamp"], unit="s").max()

print("Earliest rating timestamp:", ratings_time_min)
print("Latest rating timestamp:", ratings_time_max)

Earliest rating timestamp: 1996-03-29 18:36:55
Latest rating timestamp: 2018-09-24 14:27:30


## 4. Data Representation

From the course lectures, this dataset is **record-based data** because each row is a record describing an object or interaction.

For this project:
- A rating record represents a user-item interaction.
- The `userId` identifies the user.
- The `movieId` identifies the movie.
- The `rating` is the user's explicit preference score.
- The `timestamp` records when the rating was made.

For collaborative filtering, the main representation will later become a **user-item matrix**:
- rows = users
- columns = movies
- values = ratings

## 5. Data Quality Checks

The data quality checks below follow the course discussion of common raw data problems:
- missing values,
- duplicate records,
- noisy or invalid values,
- inconsistent identifiers,
- sparsity.

These checks help determine whether the data is ready for collaborative filtering.

In [7]:
missing_summary = pd.DataFrame({
    "ratings_missing": ratings.isna().sum(),
}).T

display("Missing values in ratings.csv")
display(ratings.isna().sum())

display("Missing values in movies.csv")
display(movies.isna().sum())

display("Missing values in tags.csv")
display(tags.isna().sum())

display("Missing values in links.csv")
display(links.isna().sum())

'Missing values in ratings.csv'

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

'Missing values in movies.csv'

movieId    0
title      0
genres     0
dtype: int64

'Missing values in tags.csv'

userId       0
movieId      0
tag          0
timestamp    0
dtype: int64

'Missing values in links.csv'

movieId    0
imdbId     0
tmdbId     8
dtype: int64

In [8]:
duplicate_summary = pd.DataFrame({
    "File": ["ratings", "movies", "tags", "links"],
    "Duplicate rows": [
        ratings.duplicated().sum(),
        movies.duplicated().sum(),
        tags.duplicated().sum(),
        links.duplicated().sum()
    ]
})

display(duplicate_summary)

,File,Duplicate rows
0,ratings,0
1,movies,0
2,tags,0
3,links,0


In [9]:
print("Minimum rating:", ratings["rating"].min())
print("Maximum rating:", ratings["rating"].max())
print("Mean rating:", ratings["rating"].mean())
print("Median rating:", ratings["rating"].median())

display(ratings["rating"].value_counts().sort_index())

Minimum rating: 0.5
Maximum rating: 5.0
Mean rating: 3.501556983616962
Median rating: 3.5


rating
0.5     1370
1.0     2811
1.5     1791
2.0     7551
2.5     5550
3.0    20047
3.5    13136
4.0    26818
4.5     8551
5.0    13211
Name: count, dtype: int64

In [10]:

# Ratings should be from 0.5 to 5.0 in half-star increments.
valid_rating_values = np.arange(0.5, 5.0 + 0.5, 0.5)
invalid_ratings = ratings[~ratings["rating"].isin(valid_rating_values)]

print("Number of invalid rating values:", len(invalid_ratings))
display(invalid_ratings.head())

Number of invalid rating values: 0


,userId,movieId,rating,timestamp


### Data Quality Findings

Based on the checks above:

1. The `ratings.csv`, `movies.csv`, and `tags.csv` files do not contain missing values.
2. The `links.csv` file contains a small number of missing `tmdbId` values, but this does not affect collaborative filtering because `links.csv` is not required for the main model.
3. No duplicate rows were found.
4. Ratings are valid and fall within the expected 0.5 to 5.0 scale.
5. The dataset is suitable for a recommender system because it contains explicit user ratings.

## 6. Identifier Consistency

The project depends on consistent user IDs and movie IDs.

The MovieLens README states that movie IDs are consistent across `ratings.csv`, `movies.csv`, `tags.csv`, and `links.csv`.

The checks below confirm whether every rated movie appears in the movie metadata table.

In [11]:
rated_movie_ids = set(ratings["movieId"].unique())
movie_metadata_ids = set(movies["movieId"].unique())

rated_movies_missing_metadata = rated_movie_ids - movie_metadata_ids
movies_without_ratings = movie_metadata_ids - rated_movie_ids

print("Rated movies missing from movies.csv:", len(rated_movies_missing_metadata))
print("Movies in movies.csv without ratings:", len(movies_without_ratings))

Rated movies missing from movies.csv: 0
Movies in movies.csv without ratings: 18


In [12]:
if len(movies_without_ratings) > 0:
    display(movies[movies["movieId"].isin(movies_without_ratings)].head(10))

,movieId,title,genres
816,1076,"Innocents, The (1961)",Drama|Horror|Thriller
2211,2939,Niagara (1953),Drama|Thriller
2499,3338,For All Mankind (1989),Documentary
2587,3456,"Color of Paradise, The (Rang-e khoda) (1999)",Drama
3118,4194,I Know Where I'm Going! (1945),Drama|Romance|War
4037,5721,"Chosen, The (1981)",Drama
4506,6668,"Road Home, The (Wo de fu qin mu qin) (1999)",Drama|Romance
4598,6849,Scrooge (1970),Drama|Fantasy|Musical
4704,7020,Proof (1991),Comedy|Drama|Romance
5020,7792,"Parallax View, The (1974)",Thriller


### Identifier Consistency Finding

Every movie that appears in `ratings.csv` also appears in `movies.csv`, which means the rating records can safely be joined with movie titles and genres.

There are a small number of movies in `movies.csv` that do not have ratings in `ratings.csv`. These are not useful for collaborative filtering because there is no user preference information for them.

## 7. Ratings Distribution and Sparsity

Collaborative filtering datasets are usually sparse because each user rates only a small fraction of all available items.

This section checks:
- number of ratings per user,
- number of ratings per movie,
- density of the user-item matrix.

In [13]:
ratings_per_user = ratings.groupby("userId").size()
ratings_per_movie = ratings.groupby("movieId").size()

display(ratings_per_user.describe().to_frame("ratings_per_user"))
display(ratings_per_movie.describe().to_frame("ratings_per_movie"))

print("Minimum ratings per user:", ratings_per_user.min())
print("Maximum ratings by one user:", ratings_per_user.max())
print("Minimum ratings per rated movie:", ratings_per_movie.min())
print("Maximum ratings for one movie:", ratings_per_movie.max())

,ratings_per_user
count,610.000000
mean,165.304918
std,269.480584
min,20.000000
25%,35.000000
50%,70.500000
75%,168.000000
max,2698.000000


,ratings_per_movie
count,9724.000000
mean,10.369807
std,22.401005
min,1.000000
25%,1.000000
50%,3.000000
75%,9.000000
max,329.000000


Minimum ratings per user: 20
Maximum ratings by one user: 2698
Minimum ratings per rated movie: 1
Maximum ratings for one movie: 329


In [14]:
num_users = ratings["userId"].nunique()
num_rated_movies = ratings["movieId"].nunique()
num_observed_ratings = len(ratings)
possible_user_movie_pairs = num_users * num_rated_movies

density = num_observed_ratings / possible_user_movie_pairs
sparsity = 1 - density

print("Number of users:", num_users)
print("Number of rated movies:", num_rated_movies)
print("Observed ratings:", num_observed_ratings)
print("Possible user-movie pairs:", possible_user_movie_pairs)
print("Matrix density:", round(density, 4))
print("Matrix sparsity:", round(sparsity, 4))

Number of users: 610
Number of rated movies: 9724
Observed ratings: 100836
Possible user-movie pairs: 5931640
Matrix density: 0.017
Matrix sparsity: 0.983


### Sparsity Finding

The user-item matrix is highly sparse. This means most users have rated only a small portion of the available movies.

This is normal for recommender systems and is the reason collaborative filtering is needed: the model estimates missing ratings based on similarity between users or similarity between items.

## 8. Movie Genre Preprocessing

The project directions recommend incorporating additional movie information, such as genre.

The `movies.csv` file stores genres as a pipe-separated string, such as:

```text
Action|Adventure|Sci-Fi
```

The following preprocessing step converts these genre strings into separate binary genre columns.
This creates additional metadata that can be used later to improve or explain recommendations.

In [15]:
# Split genres into binary indicator columns
genre_dummies = movies["genres"].str.get_dummies(sep="|")

movies_processed = pd.concat(
    [movies[["movieId", "title", "genres"]], genre_dummies],
    axis=1
)

print("Number of unique genre columns:", genre_dummies.shape[1])
print("Genre columns:")
print(list(genre_dummies.columns))

display(movies_processed.head())

Number of unique genre columns: 20
Genre columns:
['(no genres listed)', 'Action', 'Adventure', 'Animation', 'Children', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'IMAX', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']


,movieId,title,genres,(no genres listed),Action,Adventure,Animation,Children,Comedy,Crime,...,Film-Noir,Horror,IMAX,Musical,Mystery,Romance,Sci-Fi,Thriller,War,Western
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,0,0,1,1,1,1,0,...,0,0,0,0,0,0,0,0,0,0
1,2,Jumanji (1995),Adventure|Children|Fantasy,0,0,1,0,1,0,0,...,0,0,0,0,0,0,0,0,0,0
2,3,Grumpier Old Men (1995),Comedy|Romance,0,0,0,0,0,1,0,...,0,0,0,0,0,1,0,0,0,0
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance,0,0,0,0,0,1,0,...,0,0,0,0,0,1,0,0,0,0
4,5,Father of the Bride Part II (1995),Comedy,0,0,0,0,0,1,0,...,0,0,0,0,0,0,0,0,0,0


In [16]:
genre_counts = genre_dummies.sum().sort_values(ascending=False)
display(genre_counts.to_frame("movie_count"))

,movie_count
Drama,4361
Comedy,3756
Thriller,1894
Action,1828
Romance,1596
Adventure,1263
Crime,1199
Sci-Fi,980
Horror,978
Fantasy,779


### Genre Preprocessing Finding

Genres were successfully converted into binary features. These features preserve additional movie information and may later be used to improve or explain movie recommendations.

For the core collaborative filtering model, ratings remain the main input. Genres are kept as useful movie metadata.

## 9. Prepare Cleaned Tables for Storage and Modeling

The following cleaned tables are created:

1. `ratings_clean`
   - userId
   - movieId
   - rating
   - timestamp
   - rating_datetime

2. `movies_clean`
   - movieId
   - title
   - genres

3. `movies_genres`
   - movieId
   - title
   - genres
   - binary genre columns

4. `ratings_with_movies`
   - rating data merged with movie title and genre information

In [17]:
ratings_clean = ratings.copy()
ratings_clean["rating_datetime"] = pd.to_datetime(ratings_clean["timestamp"], unit="s")

movies_clean = movies.copy()

ratings_with_movies = ratings_clean.merge(
    movies_clean,
    on="movieId",
    how="left"
)

print("ratings_clean shape:", ratings_clean.shape)
print("movies_clean shape:", movies_clean.shape)
print("movies_processed shape:", movies_processed.shape)
print("ratings_with_movies shape:", ratings_with_movies.shape)

display(ratings_with_movies.head())

ratings_clean shape: (100836, 5)
movies_clean shape: (9742, 3)
movies_processed shape: (9742, 23)
ratings_with_movies shape: (100836, 7)


,userId,movieId,rating,timestamp,rating_datetime,title,genres
0,1,1,4.0,964982703,2000-07-30 18:45:03,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,1,3,4.0,964981247,2000-07-30 18:20:47,Grumpier Old Men (1995),Comedy|Romance
2,1,6,4.0,964982224,2000-07-30 18:37:04,Heat (1995),Action|Crime|Thriller
3,1,47,5.0,964983815,2000-07-30 19:03:35,Seven (a.k.a. Se7en) (1995),Mystery|Thriller
4,1,50,5.0,964982931,2000-07-30 18:48:51,"Usual Suspects, The (1995)",Crime|Mystery|Thriller


In [18]:

assert len(ratings_with_movies) == len(ratings_clean)


missing_titles_after_merge = ratings_with_movies["title"].isna().sum()
print("Missing movie titles after merge:", missing_titles_after_merge)

Missing movie titles after merge: 0


## 10. Create User-Item Matrix

This matrix is the main input for collaborative filtering in Part 2.

Rows represent users, columns represent movies, and values represent ratings.

Missing values mean the user has not rated that movie.

In [ ]:
user_item_matrix = ratings_clean.pivot_table(
    index="userId",
    columns="movieId",
    values="rating"
)

print("User-item matrix shape:", user_item_matrix.shape)
display(user_item_matrix.head())

In [19]:
# Save cleaned/preprocessed files for repeatability
ratings_clean.to_csv(DATABASE_DIR / "ratings_clean.csv", index=False)
movies_clean.to_csv(DATABASE_DIR / "movies_clean.csv", index=False)
movies_processed.to_csv(DATABASE_DIR / "movies_genres.csv", index=False)
ratings_with_movies.to_csv(DATABASE_DIR / "ratings_with_movies.csv", index=False)

print("Saved cleaned CSV files to:", DATABASE_DIR)

Saved cleaned CSV files to: /Users/mehrshad/Documents/CSE482_Project/database


## 11. Store the Data in a Local MySQL Database

Part 1 of the project asks that the data be stored in a database.

For this project, the cleaned data will be stored in a **local MySQL database** named:

```text
cse482_movielens
```

This keeps the project aligned with the course database material while avoiding connection issues with the university server.

### Tables to store

| Table | Purpose |
|---|---|
| `ratings` | User-movie rating records |
| `movies` | Movie title and genre metadata |
| `movies_genres` | Movie metadata with one-hot encoded genres |
| `ratings_with_movies` | Merged ratings and movie metadata |

Before running the MySQL cells below, make sure:
1. MySQL is installed and running locally.
2. You have a local MySQL username and password.
3. The Python packages are installed:
   - `sqlalchemy`
   - `pymysql`



In [20]:
%pip install sqlalchemy pymysql

Note: you may need to restart the kernel to use updated packages.


In [22]:
from sqlalchemy import create_engine, text

engine = create_engine(
    "mysql+pymysql://cse482_user:cse482_password@localhost:3307/cse482_movielens"
)

with engine.connect() as conn:
    result = conn.execute(text("SELECT 1;"))
    print("Connected:", result.fetchone())

Connected: (1,)


In [23]:
from sqlalchemy import create_engine, text
import pandas as pd

MYSQL_USER = "cse482_user"
MYSQL_PASSWORD = "cse482_password"
MYSQL_HOST = "localhost"
MYSQL_PORT = 3307
MYSQL_DATABASE = "cse482_movielens"

db_engine = create_engine(
    f"mysql+pymysql://{MYSQL_USER}:{MYSQL_PASSWORD}@{MYSQL_HOST}:{MYSQL_PORT}/{MYSQL_DATABASE}",
    future=True
)

print("Database engine created.")

Database engine created.


In [24]:
with db_engine.connect() as conn:

    result = conn.execute(text("SELECT 1;"))

    print("Connected:", result.fetchone())

Connected: (1,)


In [25]:
ratings_clean.to_sql(
    "ratings",
    con=db_engine,
    if_exists="replace",
    index=False
)

movies_clean.to_sql(
    "movies",
    con=db_engine,
    if_exists="replace",
    index=False
)

movies_processed.to_sql(
    "movies_genres",
    con=db_engine,
    if_exists="replace",
    index=False
)

ratings_with_movies.to_sql(
    "ratings_with_movies",
    con=db_engine,
    if_exists="replace",
    index=False
)

print("Tables successfully written to MySQL database.")

Tables successfully written to MySQL database.


In [26]:
with db_engine.connect() as conn:
    ratings_count = conn.execute(text("SELECT COUNT(*) FROM ratings;")).scalar()
    movies_count = conn.execute(text("SELECT COUNT(*) FROM movies;")).scalar()
    movies_genres_count = conn.execute(text("SELECT COUNT(*) FROM movies_genres;")).scalar()
    merged_count = conn.execute(text("SELECT COUNT(*) FROM ratings_with_movies;")).scalar()

print("ratings rows:", ratings_count)
print("movies rows:", movies_count)
print("movies_genres rows:", movies_genres_count)
print("ratings_with_movies rows:", merged_count)

ratings rows: 100836
movies rows: 9742
movies_genres rows: 9742
ratings_with_movies rows: 100836


In [27]:
with db_engine.connect() as conn:

    tables = conn.execute(text("SHOW TABLES;")).fetchall()

tables

[('movies',), ('movies_genres',), ('ratings',), ('ratings_with_movies',)]

In [28]:
pd.read_sql(

    "SELECT * FROM ratings LIMIT 5;",

    con=db_engine

)

,userId,movieId,rating,timestamp,rating_datetime
0,1,1,4.0,964982703,2000-07-30 18:45:03
1,1,3,4.0,964981247,2000-07-30 18:20:47
2,1,6,4.0,964982224,2000-07-30 18:37:04
3,1,47,5.0,964983815,2000-07-30 19:03:35
4,1,50,5.0,964982931,2000-07-30 18:48:51


In [29]:
pd.read_sql(
    "SELECT * FROM movies LIMIT 5;",
    con=db_engine
)

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


# Example Queries

In [30]:
query = """
SELECT 
    title,
    COUNT(*) AS num_ratings,
    ROUND(AVG(rating), 2) AS avg_rating
FROM ratings_with_movies
GROUP BY movieId, title
ORDER BY num_ratings DESC
LIMIT 10;
"""

top_rated_by_count = pd.read_sql(query, con=db_engine)
top_rated_by_count

,title,num_ratings,avg_rating
0,Forrest Gump (1994),329,4.16
1,"Shawshank Redemption, The (1994)",317,4.43
2,Pulp Fiction (1994),307,4.20
3,"Silence of the Lambs, The (1991)",279,4.16
4,"Matrix, The (1999)",278,4.19
5,Star Wars: Episode IV - A New Hope (1977),251,4.23
6,Jurassic Park (1993),238,3.75
7,Braveheart (1995),237,4.03
8,Terminator 2: Judgment Day (1991),224,3.97
9,Schindler's List (1993),220,4.22


In [31]:
query = """
SELECT 
    rating,
    COUNT(*) AS count
FROM ratings
GROUP BY rating
ORDER BY rating;
"""

rating_distribution_sql = pd.read_sql(query, con=db_engine)
rating_distribution_sql

,rating,count
0,0.5,1370
1,1.0,2811
2,1.5,1791
3,2.0,7551
4,2.5,5550
5,3.0,20047
6,3.5,13136
7,4.0,26818
8,4.5,8551
9,5.0,13211


In [32]:
query = """
SELECT 
    userId,
    COUNT(*) AS num_ratings
FROM ratings
GROUP BY userId
ORDER BY num_ratings DESC
LIMIT 10;
"""

top_users_by_ratings = pd.read_sql(query, con=db_engine)
top_users_by_ratings

,userId,num_ratings
0,414,2698
1,599,2478
2,474,2108
3,448,1864
4,274,1346
5,610,1302
6,68,1260
7,380,1218
8,606,1115
9,288,1055


## 12. Conclusion

Part 1 prepared the MovieLens dataset for the recommendation system project.

Completed tasks:
- Selected movie recommendation as the target domain.
- Loaded and inspected the MovieLens dataset.
- Checked data quality issues.
- Cleaned and transformed the data.
- Created genre features.
- Created the user-item matrix.
- Stored cleaned tables in a local MySQL database.

The data is now ready for Part 2, where user-based and item-based collaborative filtering algorithms will be implemented.